envoy/stream_info/filter_state.h

```diff
@@ -58,6 +58,23 @@ class FilterState {
      */
     virtual ProtobufTypes::MessagePtr serializeAsProto() const { return nullptr; }
 
+    virtual ProtobufWkt::Value serializeAsProtoValue() const {
+      ProtobufTypes::MessagePtr message = serializeAsProto();
+      ProtobufWkt::Value val;
+      if (message) {
+        Protobuf::util::JsonPrintOptions json_options;
+        json_options.preserve_proto_field_names = true;
+        std::string json;
+        const auto status = Protobuf::util::MessageToJsonString(*message, &json, json_options);
+        if (status.ok()) {
+          val.set_string_value(json);
+          return val;
+        }
```

Comment:
This logic still leave possibility for throwing an exception from `MessageToJsonString`. and it is changing the behavior, ProtobufWkt::Value is proto representation of JSON itself, so the JSON shouldn't set as string.

Reply:
I think it is ok if we define serializeAsProtoValue in all the places where serializeAsProto is defined.

Reply:
IIUC MessageToJsonString is doesn't throw.

Reply:
Why not just keep existing behavior? We can catch those exceptions and turn into null (theoretically they can throw but not realistically). This is an extension point so we never cover all places.

Reply:
Because we plan to remove all the exceptions from worker thread(#14320). And the current behavior that covert a MessagePtr to a protobuf::val through intermediate json string format doesn't seems desirable.

Reply:
We can just change the protobuf utility to not throw exceptions for those are used in worker thread. We should keep current behavior, a JSON string in ProtobufWkt::Value is not desired at all.

Reply:
The reason why we need to provide a default implementation here is that serializeAsProto is not virtual and return nullptr by default.  It is difficult to change the utility function jsonConvertValue which have a deep call stack and throw from multiple places, this requires changing the function signature of a few functions which is used frequently in the code base. From my understanding, if we override serializeAsProtoVal for all the classes whose serializeAsProto is being override, that wouldn't change the behavior in the current code base. we can leave comment here to remind the developer to override the serializeAsProtoVal function whenever they override the serializeAsProto.

Reply:
As an alternative, we can keep an extra non-throw version of jsonConvertValue without changing the current function signature, which do you prefer?

Reply:
The former solution avoid converting to json and back to proto val, which is good for performance, but as you said, would change behavior when user define their serializeAsProto method without defining serializeAsProtoVal.